File: src/unity/Runtime/Ads/MultiFullScreenAd.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EE {
    public class MultiFullScreenAd : ObserverManager<AdObserver>, IFullScreenAd, IMultiAd<IFullScreenAd> {
        private readonly List<IFullScreenAd> _items;
        private readonly ObserverHandle _handle;

        public MultiFullScreenAd() {
            _items = new List<IFullScreenAd>();
            _handle = new ObserverHandle();
        }

        public void AddItem(IFullScreenAd item) {
            _items.Add(item);
            _handle.Bind(item).AddObserver(new AdObserver {
                OnLoaded = () => DispatchEvent(observer =>
                    observer.OnLoaded?.Invoke()),

                OnLoadResult = result => DispatchEvent(observer =>
                    observer.OnLoadResult?.Invoke(result)),

                OnClicked = () => DispatchEvent(observer =>
                    observer.OnClicked?.Invoke())
            });
        }

        public void Destroy() {
            foreach (var item in _items) {
                item.Destroy();
            }
            _items.Clear();
            _handle.Clear();
        }

        public bool IsLoaded => _items.Any(item => item.IsLoaded);

        public async Task<bool> Load() {
            var tasks = new HashSet<Task<bool>>(_items.Select(async item => {
                if (item.IsLoaded) {
                    return true;
                }
                return await item.Load();
            }));
            while (tasks.Any()) {
                var next = await Task.WhenAny(tasks);
                if (next.Result) {
                    return true;
                }
                tasks.Remove(next);
            }
            return false;
        }

        public async Task<AdResult> Show() {
            var result = AdResult.Failed;
            foreach (var item in _items) {
                if (result == AdResult.Failed) {
                    // Only process if there isn't any successfully displayed ad.
                    result = await item.Show();
                }
                if (!item.IsLoaded) {
                    // Load in background.
                    _ = item.Load();
                }
            }
            return result;
        }
    }
}

File: src/ts/src/ads/MultiFullScreenAd.ts
import {
    ObserverHandle,
    ObserverManager,
} from "../core";
import { AdObserver } from "./IAd";
import {
    AdResult,
    IFullScreenAd,
} from "./IFullScreenAd";
import { IMultiAd } from "./IMultiAd";

export class MultiFullScreenAd extends ObserverManager<AdObserver> implements IFullScreenAd, IMultiAd<IFullScreenAd> {
    private readonly _items: IFullScreenAd[];
    private readonly _handle: ObserverHandle;

    public constructor() {
        super();
        this._items = [];
        this._handle = new ObserverHandle();
    }

    public addItem(item: IFullScreenAd): void {
        this._items.push(item);
        this._handle.bind(item).addObserver({
            onLoaded: () => this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded()),
            onClicked: () => this.dispatchEvent(observer => observer.onClicked && observer.onClicked()),
        });
    }

    public destroy(): void {
        for (const item of this._items) {
            item.destroy();
        }
        this._items.length = 0;
        this._handle.clear();
    }

    public get isLoaded(): boolean {
        return this._items.some(item => item.isLoaded);
    }

    public async load(): Promise<boolean> {
        let result = false;
        for (const item of this._items) {
            if (item.isLoaded) {
                continue;
            }
            if (await item.load()) {
                result = true;
            }
        }
        return result;
    }

    public async show(): Promise<AdResult> {
        let result = AdResult.Failed;
        for (const item of this._items) {
            if (result === AdResult.Failed) {
                result = await item.show();
            }
            if (!item.isLoaded) {
                // Load in background.
                const _ = item.load();
            }
        }
        return result;
    }
}